File: packages/engine/src/cpus/6502/instructions/pla.js
import { CycleType } from '..'

/**
 * PulL Accumulator (from the stack)
 *
 * @example
 * MODE          SYNTAX        HEX LEN TIM
 * Implied       PLA           $68  1   4
 *
 * Affects Flags: N Z
 */
export function* pla() {
  this.fetch()
  if (--this.pc < 0) this.pc = 0xffff
  yield CycleType.GET

  if (++this.sp > 0xff) this.sp = 0
  yield CycleType.GET

  this.acc = this.read(0x100 | this.sp)
  this.zero = this.acc === 0
  this.negative = (this.acc & 0x80) > 0
  yield CycleType.GET
}

File: packages/engine/src/cores/famibob/cpu/instructions/pla.zig
const std = @import("std");
const CPU = @import("../cpu.zig").CPU;
const Addressing = @import("../opcode.zig").Addressing;

// * PulL Accumulator (from the stack)
// *
// * MODE          SYNTAX        HEX LEN TIM
// * Implied       PLA           $68  1   4
// *
// * Affects Flags: N Z

pub fn pla(self: *CPU, addressing: Addressing) void {
    _ = addressing;
    if (self.cycle_counter == 2) {
        _ = self.fetch();
        self.pc -%= 1;
        self.next_cycle = .read;
        return;
    } else if (self.cycle_counter == 3) {
        self.sp +%= 1;
        self.next_cycle = .read;
        return;
    } else {
        self.acc = self.read(0x100 | @as(u16, self.sp));
        self.z = self.acc == 0;
        self.n = (self.acc & 0x80) > 0;
    }
}

test {
    _ = @import("pla.test.zig");
}
